src/Processors/QueryPlan/AggregatingStep.cpp

```diff
@@ -53,108 +57,149 @@ void AggregatingStep::transformPipeline(QueryPipeline & pipeline, const BuildQue
     /// Forget about current totals and extremes. They will be calculated again after aggregation if needed.
     pipeline.dropTotalsAndExtremes();
 
-    bool allow_to_use_two_level_group_by = pipeline.getNumStreams() > 1 || params.max_bytes_before_external_group_by != 0;
-    if (!allow_to_use_two_level_group_by)
+    if (!params.keys_vector.empty())
     {
-        params.group_by_two_level_threshold = 0;
-        params.group_by_two_level_threshold_bytes = 0;
+        /// ensure we have one input stream into fork processor
+        size_t grouping_sets_number = params.keys_vector.size();
+        pipeline.resize(1);
+        pipeline.addTransform(std::make_shared<ForkProcessor>(pipeline.getHeader(), grouping_sets_number));
+
+        Processors aggregating_transforms;
+        AggregatingTransformParamsPtr transform_params_ptr;
+        Processors expression_transforms;
+        // std::set<size_t>
+        for (size_t i = 0; i < grouping_sets_number; ++i)
+        {
+            params.keys = params.keys_vector[i];
+            params.keys_size = params.keys.size();
+            transform_params_ptr = std::make_shared<AggregatingTransformParams>(params, final);
+
+            aggregating_transforms.push_back(std::make_shared<AggregatingTransform>(pipeline.getHeader(), transform_params_ptr));
+
+            if (grouping_sets_number == 1)
+                break;
+
+            size_t first_column_to_add = (i == 0 ? 1 : 0);
+            auto adding_column_action = ActionsDAG::makeAddingColumnActions(pipeline.getHeader().getByPosition(params.keys_vector[first_column_to_add][0]));
```

Comment:
That's suspicious.
What if we have two sets with the same prefix?
For `(a, b), (a, c)` we add const column `a` for both, replacing normal column with default one.